posts/kotlin-like-java-but-better-1.md

```diff
@@ -0,0 +1,184 @@
+extremely useful when working with classes that are imported from a library or 
+if you want to install some new logic onto a built-in class.
+
+```kotlin
+// MutableList is Kotlin's built-in version of ArrayList
+fun MutableList<Int>.swap(index1: Int, index2: Int) {
+    val tmp = this[index1]
+    this[index1] = this[index2]
+    this[index2] = tmp
+}
+
+val list = mutableListOf(1, 2, 3)
+list.swap(0, 2) /// list is now [3, 2, 1]
+```
+
+## Coroutines
+
+Kotlin has introduced a robust 
+[coroutine](https://kotlinlang.org/docs/coroutines-overview.html) system to 
+support asynchronous programming. Coroutines are lightweight virtual 
+threads that all run on the main thread.  
+
+```kotlin
+fun main() {
+    GlobalScope.launch { // launch a new coroutine in background and continue
+        delay(1000L) // non-blocking delay for 1 second (default time unit is ms)
+        println("World!") // print after delay
+    }
+    println("Hello,") // main thread continues while coroutine is delayed
+    Thread.sleep(2000L) // block main thread for 2 seconds to keep JVM alive
+}
+```
+
+An in-depth look at all of what coroutines have to offer would probably turn
+this blog post into a short book, but they are one of the primary features 
+that allow you to utilize more modern programming principles in your Kotlin 
+code.
+
+You may also notice that Kotlin has a println() method. This is another example
+of the quality of life improvements that Kotlin provides to make the syntax 
+eaiser- no more `System.out.println()`!
+
+# How to Get Started
+If you're ready to get started with Kotlin, there are tons of great resources 
+to begin learning Kotlin.
+- [Getting Started With Kotlin](https://kotlinlang.org/docs/getting-started.html)
+- [Kotlin Koans](https://play.kotlinlang.org/koans/overview) - A set of 
+exercises to introduce someone familiar with Java to Kotlin syntax and
+patterns  
+- [Spring Boot Kotlin Tutorial](https://spring.io/guides/tutorials/spring-boot-kotlin/) - Get started creating a Spring Boot project completely in Kotlin.
+
+JetBrains describes Kotlin as "a modern but already mature programming
+language aimed to make developers happier", and I completely agree with this
+assessment. As someone with a Java background, Kotlin provides me with
+a very similar coding experience to Java, but with changes that consistently
+improve the development experience and allow me to focus on solving problems.
+
+Spend an hour or two going through the Kotlin Koans, and I guarantee you will 
+also begin noticing these small differences. Once you're ready to begin working
+with Kotlin, you can write projects from scratch in Kotlin, or utilize Kotlin's
+deep interoperability with Java. As long as you're working with Java 6 or 
+higher, it's as simple as adding the Kotlin compiler to your project and 
+beginning to write Kotlin classes. 
+
+Check out part two of this blog coming soon to walk through converting an 
```

Comment:
Part of two of this blog will walk through converting an existing Spring Boot application from Java to Kotlin.  Stay tuned!